python/ray/util/client/server/proxier.py

```diff
@@ -79,39 +94,81 @@ def _get_unused_port(self) -> int:
                 return port
         raise RuntimeError("Unable to succeed in selecting a random port.")
 
-    def start_specific_server(self, client_id) -> None:
+    def _get_session_dir(self) -> str:
+        """
+        Gets the session_dir of this running Ray session. This usually
+        looks like /tmp/ray/session_<timestamp>.
+        """
+        if self._session_dir:
+            return self._session_dir
+        connection_tuple = ray.init(address=self.redis_address)
+        ray.shutdown()
+        self._session_dir = connection_tuple["session_dir"]
+        return self._session_dir
+
+    def start_specific_server(self, client_id: str,
+                              job_config: JobConfig) -> None:
         """
         Start up a RayClient Server for an incoming client to
         communicate with.
         """
-        port = self._get_unused_port()
-        specific_server = SpecificServer(
-            port=port,
-            process_handle=start_ray_client_server(
-                self.redis_address,
-                port,
-                fate_share=self.fate_share,
-                server_type="specific-server"),
-            channel=grpc.insecure_channel(
-                f"localhost:{port}", options=GRPC_OPTIONS))
         with self.server_lock:
+            port = self._get_unused_port()
+            handle_ready = futures.Future()
+            specific_server = SpecificServer(
+                port=port,
+                process_handle_future=handle_ready,
+                channel=grpc.insecure_channel(
+                    f"localhost:{port}", options=GRPC_OPTIONS))
             self.servers[client_id] = specific_server
 
-    def get_channel(self, client_id: str) -> Optional["grpc._channel.Channel"]:
-        """
-        Find the gRPC Channel for the given client_id
-        """
-        client = None
+        serialized_runtime_env = job_config.get_serialized_runtime_env()
+
+        proc = start_ray_client_server(
+            self.redis_address,
+            port,
+            fate_share=self.fate_share,
+            server_type="specific-server",
+            serialized_runtime_env=serialized_runtime_env,
+            session_dir=self._get_session_dir())
+        # Wait for proc to reach 'client server'
+        psutil_proc = psutil.Process(proc.process.pid)
+        while True:
+            cmd = psutil_proc.cmdline()
+            if len(cmd) > 3 and cmd[2] == "ray.util.client.server":
+                break
+            logger.error(
+                "Waiting for Process to reach the actual client server.")
+            time.sleep(0.5)
+        logger.error(f"PID {proc.process.pid}")
+        handle_ready.set_result(proc)
```

Comment:
what is this doing?

Reply:
Wait for the process being run transitions from the shim process to the actual RayClient Server.